### PR TITLE
GPU support for gguf.

### PR DESCRIPTION
**Pull request by Sixzero**
A working GPU support which reproduces the CPU solution results.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ authors = ["Lukas Mayrhofer <lukas.mayrhofer@tum.de>"]
 version = "0.1.0"
 
 [deps]
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
```
Comment (cpfiffer) [inline Project.toml:7]: I was going to suggest compat bounds for this but it seems Llama2.jl doesn't have them yet. Probably a longer term issue for now.

File: src/inference.jl [hunk 2]
```diff
@@ -96,6 +96,10 @@ RunState(c::ModelConfig) = RunState(;
     logits         = zeros(Float32, c.vocab_size),
     kvcache_layers = [KVCache(c.dim ÷ c.n_heads, c.n_kv_heads, c.seq_len) for _ in 1:c.n_layers],
 )
+get_run_state(model::LanguageModel) = RunState(model.config)
+
+# A nicer silu could be? silu(x) = x*σ(x)
+silu(x) = x * (1f0 / (1f0 + exp(-x)))
```
Comment (cpfiffer) [inline src/inference.jl:102]: This should probably be 

```suggestion
silu(x::T) where {T<:Real} = x * (one(T) / (one(T) + exp(-x)))
```

because this will promote `Float16` to `Float32`:

```julia-repl
julia> @code_warntype silu(Float16(1))
MethodInstance for silu(::Float16)
  from silu(x) @ Main REPL[1]:2
Arguments
  #self#::Core.Const(silu)
  x::Float16
Body::Float32
1 ─ %1 = -x::Float16
│   %2 = Main.exp(%1)::Float16
│   %3 = (1.0f0 + %2)::Float32
│   %4 = (1.0f0 / %3)::Float32
│   %5 = (x * %4)::Float32
└──      return %5
```

File: src/inference.jl [hunk 3]
```diff
@@ -326,7 +332,8 @@ function sample(
                 # apply softmax to the logits to get the probabilities for next token
                 softmax!(state.logits)
                 # sample from this distribution to get the next token
-                next = wsample(1:config.vocab_size, state.logits)
+                logits = Array(state.logits) # copy to CPU since wsample is not supported on GPU
```
Comment (cpfiffer) [inline src/inference.jl:335]: Interesting -- probably a good place to pick up some speed. It looks like someone wrote about this [here](https://tianjun.me/essays/Categorical_Sampling_on_GPU_with_Julia/) but it's probably not important for the moment.

File: src/inferenceCUDA.jl [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+  # final rmsnorm
+  rmsnorm!(x, x, weights.rms_final_weight)
+
+  # classifier into logits
+  matmul!(s.logits, weights.output_weight, x)
+  
+  end
+  return nothing
+end
```
Comment (cpfiffer) [inline src/inferenceCUDA.jl:349]:
```suggestion
end

```

Comment (cpfiffer) [reply src/inferenceCUDA.jl:349]: (adding a newline at the end of the file)

File: src/matmul.jl [hunk 5]
```diff
@@ -11,7 +10,7 @@ function matmul!(
 end
 
 function matmul!(
-    y::AbstractVector{Float32},
+    y::Vector{Float32},
```
Comment (cpfiffer) [inline src/matmul.jl:13]: This seems relatively strict -- presumably this is to catch non-CUDA vectors?

File: test/test.jl [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+using CUDA
+using Llama2: to_cuda
+model_cu = nothing
+# we need to free cuda memory before loading a new model
+CUDA.reclaim(); GC.gc(true); #CUDA.memory_status()
+model_cu = to_cuda(model);
+
+#%%
+# GPU sample
+res = sample(model_cu, "Tim was happy."; temperature = 0.0f0, max_seq_len=10)
```
Comment (cpfiffer) [inline test/test.jl:15]: For this test file, could we add like an `@test` call that matches one of the CPU-generated texts?

File: Project.toml (hunk 1 above)
Comment (Sixzero) [reply Project.toml:7]: Actually we won't need CUDA dependency anymore, since KernelAbstraction works with "any" GPU compatible array.

File: src/inference.jl (hunk 2 above)
Comment (Sixzero) [reply src/inference.jl]: It is now going to be this: `silu(x) = x*σ(x)` 
I am not sure if it is the same. :thinking:

File: src/inference.jl (hunk 3 above)
Comment (Sixzero) [reply src/inference.jl:248]: Yeah, worth noting that one, because the `logits` is a 128000+ length vector.

File: src/inferenceCUDA.jl (hunk 4 above)
Comment (Sixzero) [reply src/inferenceCUDA.jl]: Okay, going to keep this pattern. :)

File: src/matmul.jl (hunk 5 above)
Comment (Sixzero) [reply src/matmul.jl]: Yeah it was actually the reason, but now everything is using kernels, even the CPU branches with KA, so nothing used `matmul.jl` so it got deleted.

File: test/test.jl (hunk 6 above)
Comment (Sixzero) [reply test/test.jl]: Great idea! Going to add some test cases!

Comment (Sixzero) [reply test/test.jl]: After more carefully looking at this, I guess it is not that easy to test, as models with different versions give different results. The KA kernel solution for CPU and the same kernels on GPU are not giving the same token after ~40 tokens, and I guess the old CPU code gave not exactly the same results... but AFAI remember the first 10-20 tokens were pretty similar to what the current solution produces.

**Comment by Sixzero** [comment]
I kept the CUDA.jl kernels in the test folder, because I am curious what is causing the KA kernel perform slower by quite huge margin. I guess for later on we could have some tests on what solution gives what speed, so people can catch up and see different ideas and try new ones.

**Comment by Sixzero** [comment]
I wonder if it would be worth squashing the commits into one bigger, so people can see better what has changed, because there was a quite big "sidejourney" with the CUDA.jl kernels.

**Review by just-cameron — COMMENTED** [review]
I'm noticing that there's not a lot of tests here, mostly just smoke tests. When you have a second, could you add some `@test` statements to check for expected results? I tried working through where to add them but I couldn't quite get a sense of expected output from some of these functions.

Also -- marvelous done on the improvements, all of these seem quite reasonable.

File: src/inference.jl [hunk 7]
```diff
@@ -24,25 +24,25 @@ end
...
-    token_embedding_table::Matrix # (dim, vocab_size)
-    layers::Vector{TransformerLayerWeights}
+    token_embedding_table::AbstractMatrix # (dim, vocab_size)
+    layers::Vector{TransformerLayerWeights} # NTuple{TransformerLayerWeights} # but not every layer is the same type!
     # final rmsnorm
-    rms_final_weight::Vector{Float32} # (dim,)
-    output_weight::Matrix # (dim, vocab_size)
+    rms_final_weight::AbstractVector # (dim,) # Float32
+    output_weight::AbstractMatrix # (dim, vocab_size)
 end
```
Comment (cpfiffer) [inline src/inference.jl:46]: Since `TransformerLayerWeights` is a parametric type, it might be useful here to also do

```julia
@kwdef struct TransformerWeights{T<:TransformerLayerWeights}
    token_embedding_table::AbstractMatrix # (dim, vocab_size)
    layers::Vector{T} # NTuple{TransformerLayerWeights} # but not every layer is the same type!
    # final rmsnorm
    rms_final_weight::AbstractVector # (dim,) # Float32
    output_weight::AbstractMatrix # (dim, vocab_size)
end
...

File: src/inference.jl [hunk 8]
```diff
@@ -58,128 +58,51 @@ function Base.show(io::IO, mime::MIME"text/plain", model::LanguageModel)
...
 
 struct KVCache
-    key_cache::Array{Float32,3}   # (head_size, n_heads, seq_len)
-    value_cache::Array{Float32,3} # (seq_len, head_size, n_heads)
+    key_cache::AbstractArray{Float32, 3}   # (head_size, n_heads, seq_len), {Float32,3}
+    value_cache::AbstractArray{Float32, 3} # (seq_len, head_size, n_heads), {Float32,3}
```
Comment (cpfiffer) [inline src/inference.jl:62]: Same parametric type thing here, though I think you might be able to do

```julia
struct KVCache{A<:AbstractArray{Float32, 3}}
    key_cache::A   # (head_size, n_heads, seq_len), {Float32,3}
    value_cache:: # (seq_len, head_size, n_heads), {Float32,3}
```

File: src/inference.jl [hunk 9]
```diff
@@ -58,128 +58,51 @@ function Base.show(io::IO, mime::MIME"text/plain", model::LanguageModel)
...
+    hb::AbstractVector{Float32}     # buffer for hidden dimension in the ffn (hidden_dim,)
+    hb2::AbstractVector{Float32}    # buffer for hidden dimension in the ffn (hidden_dim,)
+    q::AbstractVector{Float32}      # query (dim,)
+    k::AbstractVector{Float32}      # key (dim,)
+    v::AbstractVector{Float32}      # value (dim,)
+    att::AbstractVector{Float32}    # buffer for scores/attention values (seq_len * n_heads,)
+    logits::AbstractVector{Float32} # output logits
     # kv cache
     kvcache_layers::Vector{KVCache}
 end
```
Comment (cpfiffer) [inline src/inference.jl:84]: Same parametric type here, can do `RunState{V<:AbstractVector{Float32}}`

File: src/inference.jl (hunk 3 above)
Comment (cpfiffer) [reply src/inference.jl:247]: Let's flag it for later and do a separate PR on it. Do you think there's a good upstream place we could put this? Like a kernel library somewhere.

File: src/quantization/q6.jl [hunk 10]
```diff
@@ -90,33 +90,74 @@ function quantize!(y::Vector{block_q6_K}, x::Vector{Float32})
...
+#                 y[QK_K*(i-1) + 128*(n-1) + l + 32] = d * (Int16(scales[is + 2]) * q2)
+#                 y[QK_K*(i-1) + 128*(n-1) + l + 64] = d * (Int16(scales[is + 4]) * q3)
+#                 y[QK_K*(i-1) + 128*(n-1) + l + 96] = d * (Int16(scales[is + 6]) * q4)
+#             end
+#         end
+#     end
+
+#     return y
+# end
+
```
Comment (cpfiffer) [inline src/quantization/q6.jl:123]: Can probably remove this stuff if it's not needed for now.
```suggestion
```

**Review by just-cameron — COMMENTED** [review]
More notes! Great work in here, really love it.

File: src/inference.jl [hunk 11]
```diff
@@ -24,25 +24,25 @@ end
...
+    wk::AbstractMatrix # (dim, dim)
+    wv::AbstractMatrix # (dim, dim) # different quantization usually
+    wo::AbstractMatrix # (dim, dim)
     # weights for ffn
-    w1::Matrix # (dim, hidden_dim)
-    w2::Matrix # (hidden_dim, dim)
-    w3::Matrix # (dim, hidden_dim)
+    w1::AbstractMatrix # (dim, hidden_dim)
+    w2::AbstractMatrix # (hidden_dim, dim) # different quantization usually
+    w3::AbstractMatrix # (dim, hidden_dim)
```
[the remark this answers is not in the document]
Comment (Sixzero) [reply src/inference.jl:37]: Could we run an MVP for this? Just to see if this really is needed? I would think you are right(I guess I have seen that), but not sure. 
Just because it would make some definitions a little bit more complex.

File: src/inference.jl (hunk 7 above)
Comment (Sixzero) [reply src/inference.jl:46]: I guess for maximalising performance layers could be `Tuple`, so every type it holds is available when we iterate over the `TransformerLayerWeights`.

File: src/inference.jl (hunk 8 above)
Comment (Sixzero) [reply src/inference.jl:62]: I would think in this case Array would be Array{Float32, 3} ... also I wonder if this even compiles? :O We can use AbstractType with {} syntax? :astonished:

File: src/quantization/q6.jl (hunk 10 above)
Comment (Sixzero) [reply src/quantization/q6.jl:123]: It might still be used for some gguf quantizations, so I would still keep it. The dequantization of token weights is what might use this. https://github.com/cafaxo/Llama2.jl/blob/3d07db9354a3108d2c5dcd8ff5dea2ea9347daf8/src/inference.jl#L198

File: src/inference.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (cpfiffer) [reply src/inference.jl:37]: I would consider it to be standard practice, and it shouldn't actually impact anything downstream. Most constructors will just work.

It's also hard to test trivial examples, we'd probably need to see benchmarks before and after. It's also entirely possible that this is a nothing change, but worth checking if you're interested.

File: src/inference.jl (hunk 8 above)
Comment (cpfiffer) [reply src/inference.jl:62]: Yes, I believe so. Abstract types can be parametric as well, like this one:

```julia
https://github.com/JuliaGraphs/Graphs.jl/blob/f97b4afd17e1cd938e53282bcb76c1b8e2ea127c/src/interface.jl#L34-L38
```

File: src/quantization/q6.jl (hunk 10 above)
Comment (cpfiffer) [reply src/quantization/q6.jl:123]: Ah okay, perfect. We should plan on spinning it out at some point as well if it becomes relevant. It may also be worth adding a brief note at the top indicating what it's for and whether it works.

File: src/inference.jl (hunk 8 above)
Comment (Sixzero) [reply src/inference.jl:62]: Ah, I know what I confused this with. ParametricTypes couldn't be used with {} symbols.

File: src/inference.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (cafaxo) [reply src/inference.jl:37]: I see no real benefit to make these fields concretely-typed. Every layer can have differently quantized weights, and we do not want to compile new specialized methods for every combination of types. I deliberately made these fields abstractly-typed, to reduce the load on the compiler. I do not see how this would improve runtime performance.

[the remark this answers is not in the document]
Comment (cpfiffer) [reply src/inference.jl:37]: I wouldn't expect these types to have a massive number of different parametric types. For example, most of these will be either a GPU array or a CPU array, all of which are unlikely to change in value type and size. If it were the case that we were passing in a significant number of different types, then [yes it would be an issue](https://docs.julialang.org/en/v1/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)).

Imo the benefit of adding parametric types is to assist the compiler in 

1. Times of type instability. Parametric types can propagate upwards in the compile chain and help resolve periods where the output type would otherwise be moderately ambiguous. In this case, the compiler would observe that it should receive a `Matrix` or a `CuArray`. Without this, there is a chance that the compiler will only be able to specialize on `AbstractArray`. I don't know for sure that this is happening, and it is probably not, but including it is a "safe" option.
2. Precise function compilations. Most of these instances are heavy, large, and well lived, so compile time is probably a small part of the cost for these types. For heavy numeric operations like this, we need the most precise possible compiled operations, which is what you get from parametric types. 

In general I'm not sure how well tested the package is for type stability, but this would be an unambiguous way to force the compiler to tilt towards highly specific, well-optimized compiled code.

Up to you -- I personally think this is a very easy fix that should impact very little of the codebase. At best it gives you free speed and a higher baseline type stability. At worse you incur a very small and highly unlikely compile time cost in the event of volatile types used to construct these types.

[the remark this answers is not in the document]
Comment (cafaxo) [reply src/inference.jl:37]: The weight arrays actually have do have different `eltype`s such as `block_q4_K`, `block_q5_K`, `block_q6_K`, etc. Having the fields in the weight structs abstractly typed just means that the matmul! calls in the inference function will do need a dynamic dispatch to identify the fully specialized kernel.
If we add type parameters to the `TransformerLayerWeights`, then the layers have to be stored in a heterogeneous array which also causes dynamic dispatches.

I agree though that `RunState` and `KVCache` should be parametrized -- since we are not running into the issues with the varying `eltype`s there.

[the remark this answers is not in the document]
Comment (cpfiffer) [reply src/inference.jl:37]: Sounds fair to me -- to clarify, that means this should be able to handle at most ~7 different matrix types and a two for the `rms` fields. This doesn't sound like too much to me.

One additional question -- how much do we actually expect those eltypes to change between constructions? My intuition would have been that it's quite rare to use a significant number of different block types that change between models. I would imagine that a model is mostly fixed to use the same types once it is instantiated, in which case the static dispatch is optimal.

My point there is that if you have a dispatchable amount of types, then it is probably optimal to provide the compiler the option to use static dispatch, especially since `matmul!` is called so many times. If those types are changing every second or millisecond or whatever, then yes, this is a stupid idea. My sense too is that we shouldn't expect a large number of layers. Llama 3 70b has 80 layers, and many of those are likely to have the same types. 

I think the idea here is that there is a trade off. Dynamic dispatch can accumulate a lot of overhead for high-frequency calls. If you have a bajillion types you would otherwise need to compile, then the dynamic dispatch makes sense. Neural nets are fixed once created, so we get something more closely representing a static DAG.

In the extreme case where you use homogeneous types, I'd hazard a guess that it's optimal to parametrically type, because you only ever compile one hot path.

The worst case scenario is a network of layers with infinite mix-and-match combinations of weights -- this would take a long time to compile but might perform better in general.

I'll drop it for now, since I think I've mentioned my views and ultimately it's a design decision by you. Appreciate you walking through it with me!

File: src/inference.jl (hunk 7 above)
Comment (cpfiffer) [reply src/inference.jl:46]: Yeah, if @cafaxo is into it then I think tuples would be appropriate here, though it introduces a different consideration that the unwrapped tuples may be union types rather than type-stable: https://discourse.julialang.org/t/parametric-type-in-an-variable-length-tuple/9109/10?u=cpfiffer

Not sure if that's likely to happen here, that's probably a benchmarking thing for sure.

File: src/inference.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (Sixzero) [reply src/inference.jl:37]: I wouldn't think we need to argue about these. I pretty sure eventually going to have the time to benchmark both cases, bothfor compile time and runtime. There is no need to decide which would be more beneficial, we could also score how readable the 2 different code. :)

[the remark this answers is not in the document]
Comment (cpfiffer) [reply src/inference.jl:37]: I'm happy to move on. Just noting that this is a design consideration that may be worth thinking about for performance reasons.
